### ISSUE TITLE
Light level issue: torches brighter than sunlight

### ISSUE THREAD

**Issue by Ezhh** [minetest/minetest#7686]
##### Issue type
Bug report
 
##### Minetest version
Current dev, but also present in earlier versions.

##### OS / Hardware
Radeon RX 480

##### Summary
Sunlight is meant to be brighter than torches, but placing a torch, even at midday, visibly brightens the area it's placed in.
![torchvssun](https://user-images.githubusercontent.com/7035183/44945087-8ecb5980-add9-11e8-830e-623168ded1fa.gif)

Torch in this example is light_source = 13, and I have also tested in unedited MTG and had the same result.

##### Steps to reproduce
Make it day time. Place a torch!

**Comment by tenplus1** [comment]
Confirmed for 0.4.17-1 also.

**Comment by paramat** [comment]
This has been happening for a long time. Something to do with the fact that the torch is artificial light, so the light is made yellower, it seems a side effect of the code is a visual brightening. The node light level is still 15 of course.
It might be worth trying basic shaders on and off as i seem to remember those make a difference here.

**Comment by paramat** [comment]
Possibly relevant code in shaders:
https://github.com/minetest/minetest/blob/a4f41e7bfc6c2d823eab1465258617cb033a452c/client/shaders/nodes_shader/opengl_vertex.glsl#L125

**Comment by Ezhh** [comment]
Shaders on or off, it's the same. (Is basic shaders something separate?)

> The node light level is still 15 of course.

I assume you mean 13 here?

As a note, in current dev I have often found myself thinking it's not full daylight, only to check and find it is. My overall impression has been the game is too dark. I dismissed this as being a screen issue of some kind, even though I never experience the same feeling when playing stable (I'm actually still using a 0.4.16 client for normal play, so it's possible I would experience it in actual stable). Either way, once I place torches, the light level feels correct to me, so that made me think it's the sun that maybe isn't bright enough.

**Comment by Ezhh** [comment]
To compare, dev on left, 0.4.16 on right, all settings the same, time set to midday:
![cuvdev](https://user-images.githubusercontent.com/7035183/44945518-339e6480-ade3-11e8-94ec-d776959872bb.png)

Focusing in specifically on the sky colour:
![skycol](https://user-images.githubusercontent.com/7035183/44945522-3b5e0900-ade3-11e8-90d6-4c67faa52f46.png)

The game is definitely a lot darker now; however I think this is separate to the lighting issue and just makes it easier to see.

Taken from 0.4.16 stable, you can see the lighting issues is present even that long ago:
![sunvtorch2](https://user-images.githubusercontent.com/7035183/44945552-d525b600-ade3-11e8-8bcb-c91274d159d5.gif)

**Comment by Ezhh** [comment]
0.4.15 sky colour (bottom right) overlaying 0.4.16 sky colour.
![overlap](https://user-images.githubusercontent.com/7035183/44950888-16ea4700-ae4c-11e8-8250-e0c11e9c150f.png)
No change here.

**Comment by paramat** [comment]
Ok. I'll support changing sky colour back to 0.4.16 colours.
Note the horizon haze is known in sky.cpp as 'far cloudy fog thing'.

**Comment by paramat** [comment]
I have some experience working with sky colours so can self-assign unless you want to do it.

**Comment by Ezhh** [comment]
`video::SColorf skycolor_bright_normal_f = video::SColor(255, 140, 186, 250);`

This is the value of blue I see in stable and old clients, not the value I actually get while playing dev, so something is changing what we get after it is set here.

**Comment by Ezhh** [comment]
Okay, will see if I can match it.

**Comment by paramat** [comment]
I might have found the cause of this, and the fix may not involve messing with the light curve.

Displayed sky colour is `m_skycolor` as suggested here:
https://github.com/minetest/minetest/blob/a4f41e7bfc6c2d823eab1465258617cb033a452c/src/sky.cpp#L209
`m_skycolor` depends on a multiplier `m_brightness` which seems to have a max of 1.0:
https://github.com/minetest/minetest/blob/a4f41e7bfc6c2d823eab1465258617cb033a452c/src/sky.cpp#L656
`m_brightness` is determined partly by `time_brightness`:
https://github.com/minetest/minetest/blob/a4f41e7bfc6c2d823eab1465258617cb033a452c/src/sky.cpp#L597
`time_brightness` is an argument to the function:
https://github.com/minetest/minetest/blob/a4f41e7bfc6c2d823eab1465258617cb033a452c/src/sky.cpp#L597
And is passed here:
https://github.com/minetest/minetest/blob/7e21cec55bba7e5507cc84c2a53895429b812aff/src/game.cpp#L3668
And is set here:
https://github.com/minetest/minetest/blob/7e21cec55bba7e5507cc84c2a53895429b812aff/src/game.cpp#L3632
`daynight_ratio` has a max value of 1000 so `decode_light_f()` is given an argument that has a max of 1.0.
`decode_light_f()` is here:
https://github.com/minetest/minetest/blob/b298b0339c79db7f5b3873e73ff9ea0130f05a8a/src/light.h#L66
And when given an argument 1.0, multiplies it by `LIGHT_MAX` which is 14, adds 0.5, rounds to an integer, result 14. It then gets the visual brightness of light level 14, which since #6696 is < 255, divides it by 255 and returns, the result is a value < 1.0 causing a darker sky.

**Comment by paramat** [comment]
```
#define LIGHT_MAX 14
#define LIGHT_SUN 15

inline float decode_light_f(float light_f)
{
	s32 i = (u32)(light_f * LIGHT_MAX + 0.5);

	if (i <= 0)
		return (float)light_decode_table[0] / 255.0;
	if (i >= LIGHT_SUN)
		return (float)light_decode_table[LIGHT_SUN] / 255.0;

	float v1 = (float)light_decode_table[i - 1] / 255.0;
	float v2 = (float)light_decode_table[i] / 255.0;
	float f0 = (float)i - 0.5;
	float f = light_f * LIGHT_MAX - f0;
	return f * v2 + (1.0 - f) * v1;
}
```
The fix may be simply:
`s32 i = (u32)(light_f * LIGHT_SUN + 0.5);`

**Comment by Fixer-007** [comment]
I knew about this "bug-feature" for some time, I believe this was even discussed somewhere...

**Comment by Ezhh** [comment]
@paramat Tested and appears to work correctly. Want me to wait on numberZero or make PR?

**Comment by numberZero** [comment]
I patched that function too quickly to notice its weirdness. Need to investigate it a bit; there seems to be UB in the case of negative argument, at least.

**Comment by numberZero** [comment]
Current | #7692 | #7693
--|--|--
![screenshot_20180902_225827](https://user-images.githubusercontent.com/7352626/44960612-588df700-af0b-11e8-8e2e-fa768119ee10.png) | ![screenshot_20180902_231250](https://user-images.githubusercontent.com/7352626/44960613-59268d80-af0b-11e8-879e-8e09b0e230e6.png) | ![screenshot_20180902_233959](https://user-images.githubusercontent.com/7352626/44960614-59268d80-af0b-11e8-8bfb-8b2ddb06c7a9.png)

**Comment by paramat** [comment]
Ezhh good to hear.

I think we need to be very careful about fixing artificial light making things brighter, as there may be a reason this happens and fixing it could cause subtle new bugs.
First can we fix dark sky as it is a simple fix? Then take some time with the other issue? So i think start by merging #7692

**Comment by paramat** [comment]
It's realistic that adding a bright artificial light source in daylight would very slightly brighten the area, however the degree to which it happens in MT isn't realistic and looks a little weird.

**Comment by random-geek** [comment]
Here is the colour value with bright daylight:
![sunlight](https://user-images.githubusercontent.com/35757396/44962434-295c9300-aed5-11e8-9c9d-fa1335360bdc.png)
And with a torch:
![torch](https://user-images.githubusercontent.com/35757396/44962436-311c3780-aed5-11e8-98a4-246725863f30.png)
Presumably the sunlight is multiplied by something like (0.96, 0.96, 1.0). The artificial light is not, so it is brighter and more warm. Personally, I think it would make more sense to multiply the artificial light by an orange value and leave the sunlight alone.
This still does not solve the main issue, though.

**Comment by HybridDog** [comment]
If torch light is dark enough in comparison to sun light, you can't see it.

**Comment by paramat** [comment]
#7693 merged, dark sky fixed.

**Comment by Ezhh** [comment]
Thank you - I would say the remaining issue is not a blocker (though is still an issue, so this should not be closed).

**Comment by paramat** [comment]
Yes removed labels, which were for the dark sky bug.

**Comment by MysticTempest** [comment]
Hi, one thing I discovered recently while trying to give torches a more natural, orangeish glow; was that modifying the "artificialLight" RGB values in the client 'nodes_shader', was the key.

Found here: https://github.com/minetest/minetest/blob/e369eb1a9b2d65629d444fb305925e0d1b859b6f/client/shaders/nodes_shader/opengl_vertex.glsl#L19

Modifying it to the following values, or something very close might be what y'all are looking for. It makes the artificialLight have a slight, but nearly invisible glow during midday's full brightness, and a normal-ish brightness at night. 

```
const vec3 artificialLight = vec3(0.97, 0.97, 0.99);
```

While further variations will obviously color the light according to the values. It's the current Minetest RGB values of "1.04", that cause the super brightness.

Tested in:
0.4.17.1
5.0.0-dev

**Comment by paramat** [comment]
That won't have effect when shaders are off though, some of the cause must be in the engine.

**Comment by MysticTempest** [comment]
Ah, whoops; you're right I missed that. I took another look, this time at the engine code. The equivalent code is here: https://github.com/minetest/minetest/blob/e369eb1a9b2d65629d444fb305925e0d1b859b6f/src/mapblock_mesh.cpp#L347

Modifying it to the same values as the ones in the 'nodes_shader' gives the same effect as the shader fix.
```
static const video::SColorf artificialColor(0.97f, 0.97f, 0.99f);
```
I compiled with those settings, and if you compare it to a shader file with the "1.04" values, the brightness differential is evident when turning on & off shaders. So, both files would need to be modified.

**Comment by paramat** [comment]
Thanks. I guess we will have to keep in mind this will make artificial light darker, and players complain already it is too dark :]

**Comment by random-geek** [comment]
@paramat 
The problem is that light is interpreted as a brightness multiplied by a "percent sunlight". Torches must be bright at night, so they have a low percent sunlight (high percentage artificial light), since the current sunlight color is, well, darkness. Unfortunately, the same logic applies during the day, so the area of light around the torch is dedicated to artificial light, not sunlight.
Using the currently available variables, there is no way to know if a light source is in the bright of day or a dark cave.

**Comment by SmallJoker** [comment]
Is this still happening in 5.4.0-dev?

**Comment by paramat** [comment]
I expect so, i doubt the code has changed.

**Comment by numberZero** [comment]
> Using the currently available variables, there is no way to know if a light source is in the bright of day or a dark cave.

That is, without #5155?

**Comment by paramat** [comment]
Perhaps correct, but this issue is far too low priority to justify such a big change.

A possible solution has been described above, but i am concerned it might make all artificial light darker, when MT already suffers from inherent darkness with many users complaining about that.
I expect this issue is less of a problem than the darkening of all artificial light.

**Comment by numberZero** [comment]
@paramat this is *yet another* issue whose reason is lack of #5155.

**Comment by rubenwardy** [comment]
It's worth doing #5155 even without this bug, it's the correct way to do lighting and makes certain algorithms easier. Although, the compat would be painful - would have been better to implement that during 5.0

**Comment by paramat** [comment]
Ok, that is reasonable. numberZero, i misinterpreted your comment.
